scapy/contrib/automotive/ecu.py

```diff
@@ -135,281 +138,434 @@ def command(self):
-
-    def _update_supported_responses(self, pkt):
-        self._unanswered_packets += PacketList([pkt])
-        answered, unanswered = self._unanswered_packets.sr()
-        for _, resp in answered:
-            ecu_resp = EcuResponse(session=self.current_session,
-                                   security_level=self.current_security_level,
-                                   responses=resp)
-
-            if ecu_resp not in self._supported_responses:
-                if self.verbose:
-                    print("[+] ", repr(ecu_resp))
-                self._supported_responses.append(ecu_resp)
-            else:
-                if self.verbose:
-                    print("[-] ", repr(ecu_resp))
-        self._unanswered_packets = unanswered
+            self.log[log_key].append((pkt.time, log_value))
+
+    def __update_supported_responses(self, pkt):
+        # type: (Packet) -> None
+        """
+        Stores a given packet as supported response, if a matching request
+        packet is found in a list of the latest unanswered packets. For
+        performance improvements, this list of unanswered packets only contains
+        a fixed number of packets, defined by the `lookahead` parameter of
+        this Ecu.
+        :param pkt: Packet to be processed.
+        """
+        self.__unanswered_packets.append(pkt)
+        reduced_plist = self.__unanswered_packets[-self.lookahead:]
+        answered, unanswered = reduced_plist.sr(lookahead=self.lookahead)
+        self.__unanswered_packets = unanswered
+
+        for req, resp in answered:
+            added = False
+            current_state = copy.copy(self.state)
+            EcuState.get_modified_ecu_state(resp, req, self.state, True)
+
+            if not self.store_supported_responses:
+                continue
+
+            for sup_resp in self.__supported_responses:
+                if resp == sup_resp.key_response:
+                    if sup_resp.states is not None and \
+                            self.state not in sup_resp.states:
+                        sup_resp.states.append(current_state)
+                    added = True
+                    break
+
+            if added:
+                continue
+
+            ecu_resp = EcuResponse(current_state, responses=resp)
+            if self.verbose:
+                print("[+] ", repr(ecu_resp))
+            self.__supported_responses.append(ecu_resp)
+
+    @staticmethod
+    def sort_key_func(resp):
+        # type: (EcuResponse) -> Tuple[bool, int, int, int]
+        """
+        This sorts responses in the following order:
+        1. Positive responses first
+        2. Lower ServiceIDs first
+        3. Less supported states first
+        4. Longer (more specific) responses first
+        :param resp: EcuResponse to be sorted
+        :return: Tuple as sort key
+        """
+        # TODO: Is there a more performant way, to get only the first byte
+        #       of a packet, without the need to build it?
```

Comment:
If you can ensure your packets always come from a dissected source, you can do `pkt.original[0]`. Otherwise, you can always try using `self_build()` to only build the current layer and not the payloads.. But this won't execute any `post_build` items

Reply:
Thank you very much for this hint. I will try the approach with `pkt.original`. This sounds sufficient.

Reply:
Actually, I can not always ensure this. I will just build the top layer, which consists only consists of one field.